typescript/src/main.ts

```diff
@@ -0,0 +1,298 @@
+
+interface TokenResponse {
+    access_token: string;
+    expires_in: number;
+    refresh_token: string;
+    scope: string;
+    token_type: string;
+}
+
+class DigitalFactoryDemo {
+	private _callbackServer: Server = null;
+	private _state: string = null
+	private _pkceVerifier: string = null
+	private _tokenPair: TokenResponse = null
+
+    private _token_url: string = null;
+    private _authorizationUrl= `${OAUTH_SERVER_URL}/authorize`;
+
+    private _redirectUri: string = null;
+
+    private _signInCompleteResolve: () => void = null;
+
+	signIn(): Promise<void> {
+		this._callbackServer = createServer(this._handleRequest.bind(this))
+		this._callbackServer.listen(CALLBACK_SERVER_PORT)
+
+        this._token_url = `${OAUTH_SERVER_URL}/token`;
+        this._redirectUri = `http://localhost:${CALLBACK_SERVER_PORT}/callback`;
+
+		this._state = this._generateState();
+		this._pkceVerifier = this._generatePKCEVerifier();
+		const query = new URLSearchParams({
+			client_id: CLIENT_ID,
+			redirect_uri: this._redirectUri,
+			scope: SCOPES,
+			state: this._state,
+			response_type: 'code',
+			code_challenge: this._generatePKCEChallenge(this._pkceVerifier),
+			code_challenge_method: 'S512',
+		});
+		const signInUrl = `${this._authorizationUrl}?${query}`;
+
+        print("Open the following URL in your browser and log in to Ultimaker Digital Factory:");
+        print("");
+        print(`    ${signInUrl}`);
+        print("");
+        return new Promise<void>((resolve, reject) => {
+            this._signInCompleteResolve = resolve;
+        });
+	}
+
+	private async _handleRequest(req: IncomingMessage, res: ServerResponse): Promise<void> {
+		if (!req.url) {
+            return;
+        }
+		if (!this._pkceVerifier) {
+            return;
+        }
+
+		const url = new URL(`http://localhost:${CALLBACK_SERVER_PORT}${req.url}`);
+
+		const code = url.searchParams.get('code');
+		if (!code) {
+            return;
+        }
+
+		const state = url.searchParams.get('state')
+		if (!state || state != this._state) {
+            return;
+        }
+
+		this._tokenPair = await this._requestAccessToken(code, this._pkceVerifier);
+
+        print(`Access token: ${JSON.stringify(this._tokenPair.access_token)}`);
+
+		res.writeHead(200);
+		res.end('Sign in finished, you can now close this window.')
+
+		// reset everything
+		this._callbackServer && this._callbackServer.close();
+		this._callbackServer = null;
+		this._state = null;
+		this._pkceVerifier = null;
+
+        this._signInCompleteResolve();
+	}
+
+	private async _requestAccessToken(code: string, pkceVerifier: string): Promise<TokenResponse> {
+		const tokenResponse = await fetch(this._token_url, {
+			method: 'POST',
+			body: new URLSearchParams({
+				client_id: CLIENT_ID,
+				redirect_uri: this._redirectUri,
+				grant_type: 'authorization_code',
+				scope: SCOPES,
+				code: code,
+				code_verifier: pkceVerifier,
+			})
+		})
+
+		return tokenResponse.json()
```

Comment:
Shall we make eslint as a dev dependency to improve the syntax consistency (e.g. always `;` at the end, the use of single and double quotes, one or two lines between code blocks?)

Reply:
I just want this to be as simple as possible, given it is example code.

I've gone through and make it all more consistent, especially the `;`.